### PR TITLE
Fix Invalid cyclic error for `type T [T]`

### LINKED ISSUE #38269: [Bug]: Invalid cyclic error for `type A [A]`
### Description

Following code sample gives `invalid cyclic type reference in '[T, T]'` compile time error.

```
type A [A];     // invalid cyclic type reference in '[T, T]'
```

### Steps to Reproduce

_No response_

### Affected Version(s)

_No response_

### OS, DB, other environment details and versions

_No response_

### Related area

-> Compilation

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### PR DESCRIPTION
**Pull request by SasinduDilshara**
Fixes #38269 

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: tests/jballerina-unit-test/src/test/resources/test-src/typedefs/tuple-type-definitions-cyclic.bal [hunk 1]
```diff
@@ -327,6 +327,26 @@ function testUnionWithCyclicTuplesHashCode() {
...
+    Q3? q3 = ();
+    Q4? q4 = ();
+    Q5? q5 = ();
+    Q6 _ = q1;
+    Q6 _ = q2;
+    Q6 _ = q3;
+    Q6 _ = q4;
+    Q6 _ = q5;
+}
+
```
Comment (KavinduZoysa) [inline tests/jballerina-unit-test/src/test/resources/test-src/typedefs/tuple-type-definitions-cyclic.bal:349]: Let's add a module level test too.

File: tests/jballerina-unit-test/src/test/resources/test-src/typedefs/tuple-type-definitions-cyclic.bal [hunk 2]
```diff
@@ -327,6 +327,26 @@ function testUnionWithCyclicTuplesHashCode() {
     "\"superType\":[\"|\", [\"|\", \"int\", \"float\"], [\"tuple\", \"never\", [\"|\", \"int\", \"string\"]]]}"); 
 }
 
+type Q1 [Q1];
```
Comment (ushirask) [inline tests/jballerina-unit-test/src/test/resources/test-src/typedefs/tuple-type-definitions-cyclic.bal:330]: Should we add test cases when a type is a tuple of a different cyclic type? i.e. `Q7 = [Q1]`

**Comment by SasinduDilshara** [comment]
This PR will wait untill the changes done in https://github.com/ballerina-platform/ballerina-lang/pull/37773.

**Comment by SasinduDilshara** [comment]
> This PR will wait untill the changes (#36538) done in #37773.

Hence mark this as a draft

**Pull request closed without merging**